tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/source_files/source_files_container_test.ts

```diff
@@ -0,0 +1,237 @@
+
+describe('Source Files Container', () => {
+  let store: MockStore<State>;
+  let dispatchSpy: jasmine.Spy;
+
+  class FakeRange {
+    constructor(
+      readonly startLineNumber: number,
+      readonly startColumn: number,
+      readonly endLineNumber: number,
+      readonly endColumn: number
+    ) {}
+  }
+
+  let loadMonacoSpy: jasmine.Spy;
+  // TODO(cais): Explore better typing by depending on 3rd-party libraries.
+  let editorSpy: jasmine.SpyObj<any>;
+  let monaco: any;
+  function setUpMonacoFakes() {
+    async function fakeLoadMonaco() {
+      monaco = {
+        editor: {
+          create: () => {
+            editorSpy = jasmine.createSpyObj('editorSpy', [
+              'deltaDecorations',
+              'revealLineInCenter',
+              'setValue',
+            ]);
+            return editorSpy;
+          },
+          ScrollType: {
+            Immediate: 1,
+            Smooth: 0,
+          },
+        },
+        Range: FakeRange,
+      };
+      loadMonacoShim.windowWithRequireAndMonaco.monaco = monaco;
+    }
+    loadMonacoSpy = spyOn(loadMonacoShim, 'loadMonaco').and.callFake(
+      fakeLoadMonaco
+    );
+  }
+
+  function tearDownMonacoFakes() {
+    if (loadMonacoShim.windowWithRequireAndMonaco.monaco !== undefined) {
+      delete loadMonacoShim.windowWithRequireAndMonaco.monaco;
+    }
+  }
+
+  beforeEach(async () => {
+    setUpMonacoFakes();
+    await TestBed.configureTestingModule({
+      declarations: [DebuggerComponent, DebuggerContainer],
+      imports: [
+        AlertsModule,
+        CommonModule,
+        ExecutionDataModule,
+        InactiveModule,
+        SourceFilesModule,
+        StackTraceModule,
+        TimelineModule,
+      ],
+      providers: [
+        provideMockStore({
+          initialState: createState(createDebuggerState()),
+        }),
+        DebuggerContainer,
+      ],
+    }).compileComponents();
+    store = TestBed.get(Store);
+    dispatchSpy = spyOn(store, 'dispatch');
+  });
+
+  afterEach(() => {
+    tearDownMonacoFakes();
+  });
+
+  it('renders no file selected when no source line is focused on', () => {
+    const fixture = TestBed.createComponent(SourceFilesContainer);
+    store.setState(createState(createDebuggerState()));
+    fixture.detectChanges();
+
+    const noFileSelectedElement = fixture.debugElement.query(
+      By.css('.no-file-selected')
+    );
+    expect(noFileSelectedElement.nativeElement.innerText).toBe(
+      '(No file selected)'
+    );
+  });
+
+  it('renders file path and editor when a file is focused on', async () => {
+    const fixture = TestBed.createComponent(SourceFilesContainer);
+    store.overrideSelector(getFocusedSourceFileContent, {
+      loadState: DataLoadState.LOADED,
+      lines: ['import tensorflow as tf', '', 'print("hello, world")'],
+    });
+    store.overrideSelector(getFocusedSourceLineSpec, {
+      host_name: 'localhost',
+      file_path: '/home/user/main.py',
+      lineno: 3,
+    });
```

Comment:
If you override the selectors before you create the fixture, you don't have to `refreshState`

Reply:
I think it is more like, `overrideSelector` **before** creating component does not require `refreshState` but all the subsequent overrides require one.